crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2316,7 +2349,8 @@ impl<'db> TypeInferenceBuilder<'db> {
     }
 
     fn module_ty_from_name(&self, module_name: &ModuleName) -> Option<Type<'db>> {
-        resolve_module(self.db, module_name).map(|module| Type::ModuleLiteral(module.file()))
+        resolve_module(self.db, module_name)
+            .map(|module| Type::ModuleLiteral(ModuleLiteralType::new(self.db, self.file, module)))
```

Comment:
Yeah that's a good point that this would make us construct a `ModuleName` every time we dereference a module.  I had considered a more complex representation for `ModuleName`, but figured it would be better to experiment with that separately instead of trying to fold it into this.  (Or better to just say this is good enough and leave it alone for now)

Reply:
I think that unless we a) are seeing a clear regression in the existing benchmark, or b) there's a very clear alternative that will definitely perform better, it's best not to spend much time on performance optimizations when we don't even know if they are meaningful in the overall performance profile. So I would suggest leaving this as-is for now.